tensorboard/uploader/auth_test.py

```diff
@@ -149,5 +155,305 @@ def test_read_invalid_json_file(self):
+        auth.authenticate_user(force_console=True)
+        self.mocked_installed_auth_flow_creator_fn.assert_not_called()
+        self.mocked_device_auth_flow.assert_called_once()
+        self.mocked_device_auth_flow.return_value.run.assert_called_once()
+
+
+class FakeHttpResponse:
+    """A fake implementation of the response from the requests library."""
+
+    def __init__(self, data: Dict, status: int = 200):
+        self.status_code = status
+        self._data = data
+
+    def json(self):
+        return self._data
+
+
+class LimitedInputDeviceAuthFlowTest(tb_test.TestCase):
+    _OAUTH_CONFIG = {
+        "client_id": "console_client_id",
+        "token_uri": "https://google.com/token",
+        "client_secret": "console_client_secret",
+    }
+
+    _SCOPES = ["email", "openid"]
+
+    _DEVICE_RESPONSE = FakeHttpResponse(
+        {
+            "device_code": "resp_device_code",
+            "verification_url": "auth.google.com/device",
+            "user_code": "resp_user_code",
+            "interval": 5,
+            "expires_in": 300,
+        }
+    )
+
+    _AUTH_GRANTED_RESPONSE = FakeHttpResponse(
+        {
+            "access_token": "some_access_token",
+            "refresh_token": "some_refresh_token",
+            "id_token": "some_id_token",
+            "expires_in": 3600,  # seconds
+        }
+    )
+
+    def setUp(self):
+        super().setUp()
+
+        self.mocked_time = self.enter_context(
+            mock.patch.object(
+                time,
+                "time",
+                # Timestamps from a fake clock.
+                # The values don't matter in most tests.
+                side_effect=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+            )
+        )
+
+        self.mocked_sleep = self.enter_context(
+            mock.patch.object(time, "sleep", autospec=True)
+        )
+
+        self.mocked_post = self.enter_context(
+            mock.patch.object(requests, "post", autospec=True)
+        )
+
+        self.flow = auth._LimitedInputDeviceAuthFlow(
+            self._OAUTH_CONFIG,
+            self._SCOPES,
+        )
+
+    def test_run__device_request_fails__raises(self):
+        self.mocked_post.return_value = FakeHttpResponse(
+            {"error": "quota exceeded"}, status=403
+        )
+
+        expected_error_msg = (
+            "There was an error while contacting Google's "
+            "authorization server. Please try again later."
+        )
+        with self.assertRaisesRegex(RuntimeError, expected_error_msg):
+            self.flow.run()
+
+    def test_run__polling__auth_pending_response__keeps_polling(self):
+        auth_pending_response = FakeHttpResponse(
+            {"error": "authorization_pending"}, status=428
+        )
+        self.mocked_post.side_effect = [
+            self._DEVICE_RESPONSE,
+            auth_pending_response,
+            self._AUTH_GRANTED_RESPONSE,
+        ]
+
+        self.flow.run()
+
+        device_params = {
```

Comment:
Maybe rename expected_device_params and expected_polling_params?

Reply:
Done.